File: app/src/main/kotlin/tech/mmarca/openvitals/ui/components/InsightCards.kt
package tech.mmarca.openvitals.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp

data class InsightStat(
    val title: String,
    val value: String,
    val unit: String,
    val icon: ImageVector,
    val accentColor: Color,
)

@Composable
fun InsightStatGrid(
    stats: List<InsightStat>,
    modifier: Modifier = Modifier,
    columns: Int = 2,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        stats.chunked(columns.coerceAtLeast(1)).forEach { rowStats ->
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                rowStats.forEach { stat ->
                    InsightStatCard(
                        stat = stat,
                        modifier = Modifier.weight(1f),
                    )
                }
            }
        }
    }
}

@Composable
private fun InsightStatCard(
    stat: InsightStat,
    modifier: Modifier = Modifier,
) {
    OpenVitalsCard(
        modifier = modifier.fillMaxWidth(),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Icon(
                imageVector = stat.icon,
                contentDescription = null,
                tint = stat.accentColor,
                modifier = Modifier.size(24.dp),
            )
            Spacer(Modifier.height(16.dp))
            Row(
                verticalAlignment = Alignment.Bottom,
                horizontalArrangement = Arrangement.Center,
            ) {
                Text(
                    text = stat.value,
                    style = MaterialTheme.typography.headlineSmall,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onSurface,
                    modifier = Modifier.alignByBaseline(),
                )
                if (stat.unit.isNotBlank()) {
                    Spacer(Modifier.width(4.dp))
                    Text(
                        text = stat.unit,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.alignByBaseline(),
                    )
                }
            }
            Spacer(Modifier.height(8.dp))
            Text(
                text = stat.title,
                style = MaterialTheme.typography.labelMedium,
                fontWeight = FontWeight.SemiBold,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
            )
        }
    }
}

File: lib/ui/components/insight_cards.dart
import 'package:flutter/material.dart';

import 'ov_card.dart';

/// Port of the Kotlin `InsightCards.kt`: the small stat cards every metric
/// detail screen lays out in a two-column grid.

/// Kotlin `InsightStat`.
@immutable
class InsightStat {
  const InsightStat({
    required this.title,
    required this.value,
    required this.unit,
    required this.icon,
    required this.accentColor,
  });

  final String title;
  final String value;
  final String unit;
  final IconData icon;
  final Color accentColor;
}

/// Kotlin `InsightStatGrid`: stats chunked into rows of [columns].
///
/// A trailing odd stat occupies one column and leaves the rest of its row empty,
/// as Compose's `Row` + `weight(1f)` does — it does not stretch to full width.
class InsightStatGrid extends StatelessWidget {
  const InsightStatGrid({
    super.key,
    required this.stats,
    this.columns = 2,
  });

  final List<InsightStat> stats;
  final int columns;

  @override
  Widget build(BuildContext context) {
    if (stats.isEmpty) return const SizedBox.shrink();
    final perRow = columns < 1 ? 1 : columns;

    final rows = <List<InsightStat>>[];
    for (var i = 0; i < stats.length; i += perRow) {
      rows.add(stats.sublist(i, (i + perRow).clamp(0, stats.length)));
    }

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: [
        for (final row in rows)
          IntrinsicHeight(
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                for (var i = 0; i < perRow; i++) ...[
                  if (i > 0) const SizedBox(width: 12),
                  Expanded(
                    child: i < row.length
                        ? _InsightStatCard(stat: row[i])
                        // Keeps the last row's cards the same width as the rest.
                        : const SizedBox.shrink(),
                  ),
                ],
              ],
            ),
          ),
      ],
    );
  }
}

class _InsightStatCard extends StatelessWidget {
  const _InsightStatCard({required this.stat});

  final InsightStat stat;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return OpenVitalsCard(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Icon(stat.icon, size: 24, color: stat.accentColor),
            const SizedBox(height: 16),
            // scaleDown instead of ellipsis: a shrunk numeral stays true where
            // a truncated one misreads, and the unit could still overflow on
            // its own at large font scales.
            FittedBox(
              fit: BoxFit.scaleDown,
              child: Row(
                mainAxisSize: MainAxisSize.min,
                crossAxisAlignment: CrossAxisAlignment.baseline,
                textBaseline: TextBaseline.alphabetic,
                children: [
                  Text(
                    stat.value,
                    style: theme.textTheme.headlineSmall?.copyWith(
                      fontWeight: FontWeight.bold,
                      color: theme.colorScheme.onSurface,
                    ),
                  ),
                  if (stat.unit.trim().isNotEmpty) ...[
                    const SizedBox(width: 4),
                    Text(
                      stat.unit,
                      style: theme.textTheme.bodyMedium
                          ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
                    ),
                  ],
                ],
              ),
            ),
            const SizedBox(height: 8),
            Text(
              stat.title,
              textAlign: TextAlign.center,
              style: theme.textTheme.labelMedium?.copyWith(
                fontWeight: FontWeight.w600,
                color: theme.colorScheme.onSurfaceVariant,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
